add new flag --initramfs-diskless-boot

HRio opened the pull request:
This flag skips running hook scripts and enables other options
needed during *diskless* initramfs-boot.

This flag *must* be used during initramfs tmpfs initial install.
The reason that this new flag is needed is that the hooks will currently
always fail as musl and /bin/sh is missing at this stage on diskless.

HRio committed:
add new flag --no-commit-hooks

This flag skips running hook scripts

This flag *must* be used during initramfs tmpfs initial install.
The reason that this new flag is needed is that the hooks will currently
always fail as musl and /bin/sh is missing at this stage on diskless.

HRio commented:
@fabled 

Tested with a patched mkinitfs package

```
> --- a/initramfs-init.in
> +++ b/initramfs-init.in
> @@ -618,7 +618,7 @@ mount -o bind /sys $sysroot/sys
>  mount -o bind /proc $sysroot/proc                                                                                                                                                                                                                                              
>  mount -o bind /dev $sysroot/dev
>  if [ -n "$ovlfiles" ]; then
> -       apk add --root $sysroot $repo_opt $apkflags $pkgs <$ovlfiles
> +       apk add --initramfs-diskless-boot --root $sysroot $repo_opt $apkflags $pkgs <$ovlfiles
>  else
>         apk add --root $sysroot $repo_opt $apkflags $pkgs
>  fi
```

From diskless target serial console:
```
> [    3.896166] EXT4-fs (mmcblk0p2): recovery complete                     
> [    3.910447] EXT4-fs (mmcblk0p2): mounted filesystem with ordered data mode. Opts: (null)
> ok.                                                                   
>  * Loading user settings from /media/mmcblk0p1/cubox.apkovl.tar.gz: ok.                          
>  * Installing packages to root filesystem: Skipping: etc/apk/commit_hooks.d/dummy pre-commit
> (1/35) Installing musl (1.1.18-r2)                                                               
> (2/35) Installing busybox (1.27.2-r7)                                                                                                  
> Executing busybox-1.27.2-r7.post-install                             

....snip....

> (35/35) Installing zlib-dev (1.2.11-r1)         
> Executing busybox-1.27.2-r7.trigger                            
> Skipping: etc/apk/commit_hooks.d/dummy post-commit                                                     
> OK: 37 MiB in 35 packages                                
> ok.                                                                              
>                                                                                           
>                                                              
>    OpenRC 0.24.1.a941ee4a0b is starting up Linux 4.9.65-1-hardened (armv7l)    

```

jirutka commented:
I don't like the option name. It does not communicate what it actually do, moreover skipping hooks may be useful in more situations than just this one. Why not just name it `--skip-hooks`?

fabled commented:
@jirutka the name is like that because I asked for it. The intent is to make the same switch set proper defaults as needed.
Though, thinking more of this, it might make sense to have individual flags for each item, and have the generic umbrella flag to be a shortcut to just enable appropriate things. Thoughts?

jirutka commented:
> ...have individual flags for each item, and have the generic umbrella flag to be a shortcut to just enable appropriate things

:+1: for individual flags.

Do you currently know about more flags that will be set by that umbrella flag?

fabled commented on src/apk.c:
```diff
@@ -217,6 +217,13 @@ static int option_parse_commit(void *ctx, struct apk_db_options *dbopts, int opt
+		break;
+	case 0x118:
+		dbopts->open_flags |= APK_OPENF_CREATE;
+		apk_flags |= APK_FORCE | APK_PROGRESS | APK_SKIP_HOOKS;
```
Let's just enable APK_FORCE and APK_NO_COMMIT_HOOKS for now.

HRio replied:
OK, fixing

fabled commented on src/apk.c:
```diff
@@ -228,6 +235,9 @@ static const struct apk_option options_commit[] = {
 	{ 0x102, "clean-protected",	"Do not create .apk-new files in configuration dirs" },
 	{ 0x111, "overlay-from-stdin",	"Read list of overlay files from stdin" },
 	{ 0x113, "no-scripts",		"Do not execute any scripts" },
+	{ 0x117, "skip-hooks",		"Skip pre/post hook scripts (but not other scripts)" },
```
Let's do `--no-commit-hooks` this in harmony with `no-scripts` and differentiates them better.

HRio replied:
agree, will fix

fabled commented on src/apk.c:
```diff
@@ -228,6 +235,9 @@ static const struct apk_option options_commit[] = {
 	{ 0x113, "no-scripts",		"Do not execute any scripts" },
+	{ 0x117, "skip-hooks",		"Skip pre/post hook scripts (but not other scripts)" },
+	{ 0x118, "initramfs-diskless-boot",
+	  "Enables options for diskelss initrams boot (e.g. skip hooks)" },
```
Typo with `diskelss`.

HRio replied:
thanks,

fabled commented on src/apk_defines.h:
```diff
@@ -77,6 +77,7 @@ extern char **apk_argv;
 #define APK_OVERLAY_FROM_STDIN	0x2000
 #define APK_NO_SCRIPTS		0x4000
 #define APK_NO_CACHE		0x8000
+#define APK_SKIP_HOOKS		0x00010000
```
rename to `APK_NO_COMMIT_HOOKS`

fabled reviewed (changes requested):
With the few changes I'm happy to pull these in. Thanks for coping with the fact we changed desired naming along the way.

fabled commented:
End result looks good. Please flatten commits to one (doing it all) or two (separating the umbrella flag to individual change).

HRio committed:
add new umbrella flag --initramfs-diskless-boot

This flag enables a group of options used during initramfs tmpfs initial install.
